beacon_node/beacon_chain/src/beacon_block_reward.rs

```diff
@@ -205,8 +205,9 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             for index in attesting_indices {
                 let index = index as usize;
                 for (flag_index, &weight) in PARTICIPATION_FLAG_WEIGHTS.iter().enumerate() {
-                    let epoch_participation =
-                        state.get_epoch_participation_mut(data.target.epoch)?;
+                    let mut epoch_participation =
+                        state.get_epoch_participation(data.target.epoch)?.clone();
```

Comment:
This is almost right, but we need to clone the epoch participation outside the loop and keep track of all the changes during the `for attestation in block.body().attestations() { .. }` loop. The intuition is that multiple attestations for a single validator could be included in a single block, and we don't want to double-reward them for the repeated attestations.

I think something like:

```rust
let mut current_epoch_participation = state.current_epoch_participation()?.clone();
let mut previous_epoch_participation = state.previous_epoch_participation()?.clone();
...
// Later
let epoch_participation = if data.target.epoch == state.current_epoch() {
    &mut current_epoch_participation
} else {
    &mut previous_epoch_participation
};
```

Reply:
makes sense, thanks for the additional context!